diff --git a/src/auth/Auth.cpp b/src/auth/Auth.cpp
index 6c9b5c6..97efa64 100644
--- a/src/auth/Auth.cpp
+++ b/src/auth/Auth.cpp
@@ -67,6 +67,7 @@ namespace SDDM {
         QString user { };
         QByteArray cookie { };
         bool autologin { false };
+        bool fingerprintlogin { false };
         bool greeter { false };
         QProcessEnvironment environment { };
         qint64 id { 0 };
@@ -225,12 +226,13 @@ namespace SDDM {
             Q_EMIT qobject_cast<Auth*>(parent())->error(child->errorString(), ERROR_INTERNAL);
         }
 
-        if (exitCode == HELPER_SUCCESS)
+        if (exitCode == HELPER_SUCCESS) {
             qDebug() << "Auth: sddm-helper exited successfully";
+            emit qobject_cast<Auth*>(parent())->finished(static_cast<Auth::HelperExitStatus>(exitCode));
+        }
+
         else
             qWarning("Auth: sddm-helper exited with %d", exitCode);
-
-        Q_EMIT qobject_cast<Auth*>(parent())->finished((Auth::HelperExitStatus)exitCode);
     }
 
     void Auth::Private::childError(QProcess::ProcessError error) {
@@ -276,6 +278,10 @@ namespace SDDM {
         return d->autologin;
     }
 
+    bool Auth::fingerprintlogin() const {
+        return d->fingerprintlogin;
+    }
+
     bool Auth::isGreeter() const
     {
         return d->greeter;
@@ -334,6 +340,12 @@ namespace SDDM {
         }
     }
 
+    void Auth::setFingerprintlogin(bool on){
+        if(on != d->fingerprintlogin){
+            d->fingerprintlogin = on;
+        }
+    }
+
     void Auth::setGreeter(bool on)
     {
         if (on != d->greeter) {
@@ -377,10 +389,17 @@ namespace SDDM {
             args << QStringLiteral("--user") << d->user;
         if (d->autologin)
             args << QStringLiteral("--autologin");
+        if (d->fingerprintlogin)
+            args << QStringLiteral("--fingerprintlogin");
         if (!d->displayServerCmd.isEmpty())
             args << QStringLiteral("--display-server") << d->displayServerCmd;
         if (d->greeter)
             args << QStringLiteral("--greeter");
+        if(d->child->state() != QProcess::NotRunning){
+            d->child->terminate();
+            d->child->waitForFinished();
+        }
+        qDebug() << "starting sddm-helper with" << args;
         d->child->start(QStringLiteral("%1/sddm-helper").arg(QStringLiteral(LIBEXEC_INSTALL_DIR)), args);
     }
 
@@ -392,7 +411,7 @@ namespace SDDM {
         d->child->terminate();
 
         // wait for finished
-        if (!d->child->waitForFinished(5000))
+        if (!d->child->waitForFinished(300))
             d->child->kill();
     }
 }
diff --git a/src/daemon/Display.cpp b/src/daemon/Display.cpp
index 4014e0d..6110796 100644
--- a/src/daemon/Display.cpp
+++ b/src/daemon/Display.cpp
@@ -43,6 +43,7 @@
 #include <sys/ioctl.h>
 #include <fcntl.h>
 
+#include <QLocalSocket>
 #include <QDBusConnection>
 #include <QDBusMessage>
 #include <QDBusReply>
@@ -149,6 +150,11 @@ namespace SDDM {
         // connect login signal
         connect(m_socketServer, &SocketServer::login, this, &Display::login);
 
+        connect(m_socketServer, &SocketServer::connected, [this](QLocalSocket* socket){
+            m_socket = socket;
+            fingerprintLogin();
+        });
+
         // connect login result signals
         connect(this, &Display::loginFailed, m_socketServer, &SocketServer::loginFailed);
         connect(this, &Display::loginSucceeded, m_socketServer, &SocketServer::loginSucceeded);
@@ -261,6 +267,35 @@ namespace SDDM {
         startSocketServerAndGreeter();
     }
 
+    bool Display::fingerprintLogin(){
+        if(mainConfig.Fingerprintlogin.User.get().isEmpty()){
+            return false;
+        }
+
+        Session::Type sessionType = Session::X11Session;
+
+        QString fingerprintSession = mainConfig.Fingerprintlogin.Session.get();
+        if(fingerprintSession.isEmpty()){
+            fingerprintSession = stateConfig.Last.Session.get();
+        }
+        if (findSessionEntry(mainConfig.X11.SessionDir.get(), fingerprintSession)) {
+            sessionType = Session::X11Session;
+        } else if (findSessionEntry(mainConfig.Wayland.SessionDir.get(), fingerprintSession)) {
+            sessionType = Session::WaylandSession;
+        } else {
+            qCritical() << "Unable to find autologin session entry" << fingerprintSession;
+            return false;
+        }
+        Session session;
+        session.setTo(sessionType, fingerprintSession);
+
+        m_auth->setFingerprintlogin(true);
+        m_fingerprintAuthActive = true;
+        startAuth(mainConfig.Fingerprintlogin.User.get(), QString(), session);
+        m_auth->setFingerprintlogin(false);
+        return true;
+    }
+
     void Display::displayServerStarted() {
         // check flag
         if (m_started)
@@ -327,6 +362,11 @@ namespace SDDM {
             return;
         }
 
+        if(password.isEmpty() && !m_auth->fingerprintlogin()){
+            qDebug() << "use fingerprint because password is empty";
+            m_auth->setFingerprintlogin(true);
+        }
+
         // authenticate
         startAuth(user, password, session);
     }
@@ -375,8 +415,14 @@ namespace SDDM {
     bool Display::startAuth(const QString &user, const QString &password, const Session &session) {
 
         if (m_auth->isActive()) {
-            qWarning() << "Existing authentication ongoing, aborting";
-            return false;
+            if (m_fingerprintAuthActive) {
+                qDebug() << "Stopping fingerprint auth to allow password auth";
+                m_auth->stop();
+                m_fingerprintAuthActive = false;
+            } else {
+                qWarning() << "Existing authentication ongoing, aborting";
+                return false;
+            }
         }
 
         m_passPhrase = password;
@@ -463,6 +509,7 @@ namespace SDDM {
     }
 
     void Display::slotAuthenticationFinished(const QString &user, bool success) {
+        m_fingerprintAuthActive = false;
         if (m_auth->autologin() && !success) {
             handleAutologinFailure();
             return;
diff --git a/src/daemon/Display.h b/src/daemon/Display.h
index 98ca006..daa55a9 100644
--- a/src/daemon/Display.h
+++ b/src/daemon/Display.h
@@ -73,6 +73,7 @@ namespace SDDM {
                    const QString &user, const QString &password,
                    const Session &session);
         bool attemptAutologin();
+        bool fingerprintLogin();
         void displayServerStarted();
 
     signals:
@@ -96,6 +97,7 @@ namespace SDDM {
 
         bool m_relogin { true };
         bool m_started { false };
+        bool m_fingerprintAuthActive { false };
 
         int m_terminalId = 0;
         int m_sessionTerminalId = 0;
